File: docsapp/com.ibm.concord.spreadsheet/src/com/ibm/concord/spreadsheet/document/model/impl/Column.java
package com.ibm.concord.spreadsheet.document.model.impl;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;

import org.apache.commons.lang.builder.ToStringBuilder;
import org.apache.commons.lang.builder.ToStringStyle;

import com.ibm.concord.spreadsheet.document.model.style.StyleManager;
import com.ibm.concord.spreadsheet.document.model.style.StyleObject;
import com.ibm.concord.spreadsheet.document.model.util.ModelHelper;
import com.ibm.concord.spreadsheet.document.model.util.ModelHelper.Position;

public class Column extends BasicModel
{
  private static final Logger LOG = Logger.getLogger(Column.class.getName());
  
  static Method getCoverInfoIndexMethod;
  static Method repeatCoverInfoMethod;
  
  private int width;

  private StyleObject style;

  private int repeatedNum;

  private Visibility visibility;

  private Sheet parent;
  
  private List<CoverInfo> coverList;

  static {
    try{
      getCoverInfoIndexMethod = CoverInfo.class.getDeclaredMethod("getRowIndex", null);
    } catch(NoSuchMethodException sme)
    {
      LOG.log(Level.WARNING, " no getRowIndex method in CoverInfo class");
    }
    
    try{
      repeatCoverInfoMethod = CoverInfo.class.getDeclaredMethod("getRowRepeatedNum", null);
    } catch(NoSuchMethodException sme)
    {
      LOG.log(Level.WARNING, " no getRowSpan method in CoverInfo class");
    }
  }
  
  public Column(Sheet parent, int id)
  {
    this.parent = parent;
    this.id = id;
    this.width = -1;
    this.style = null;
    this.visibility = null;
    this.repeatedNum = 0;
    this.coverList = new ArrayList<CoverInfo>();
  }

  public Column(int index, Column col)
  {
    if (null == col)
      return;
    this.parent = col.getParent();
    this.id = this.getIDManager().getColIdByIndex(this.parent.getId(), index, true);
    this.width = col.getWidth();
    this.visibility = col.getVisibility();
    this.repeatedNum = col.getRepeatedNum();
    this.coverList = new ArrayList<CoverInfo>();
    this.coverList.addAll(col.getCoverList());
    this.setStyle(col.getStyle());
  }

  public int getIndex()
  {
    return this.getIDManager().getColIndexById(this.parent.getId(), this.id);
  }

  public void copy(BasicModel model)
  {
    if (null == model)
      return;
    Column col = (Column) model;
    this.width = col.getWidth();
    this.visibility = col.getVisibility();
    this.repeatedNum = col.getRepeatedNum();
    this.setStyle(col.getStyle());
  }


  public int getWidth()
  {
    return this.width;
  }

  public void setWidth(int w)
  {
    this.width = w;
  }

  public int getRepeatedNum()
  {
    return this.repeatedNum;
  }

  public void setRepeatedNum(int num)
  {
    if(null != this.style)
    {
      StyleManager styleManager = this.getStyleManager();
      styleManager.changeRefCount(this.style, num-this.repeatedNum);
    }  
    this.repeatedNum = num;
  }

  public StyleObject getStyle()
  {
    return this.style;
  }

  public StyleManager getStyleManager()
  {
    return this.parent.getParent().getStyleManager();
  }
  
  public void setStyle(StyleObject style)
  {
    StyleManager styleManager = this.getStyleManager();
    if(null != this.style)
      styleManager.changeRefCount(this.style, -(this.repeatedNum+1));
    this.style = style;
    if(null != this.style)
      styleManager.changeRefCount(this.style, this.repeatedNum+1);
  }

  public Visibility getVisibility()
  {
    return visibility;
  }

  public void setVisibility(Visibility visibility)
  {
    this.visibility = visibility;
  }


  public boolean isContainStyle()
  {
    if(this.width != -1 || (null != this.style && 
        !this.style.equals(this.getStyleManager().getDefaultCellStyle())))
      return true;
    return false;
  }
  
  public boolean isVisible()
  {
    if(null == this.visibility || Visibility.VISIBLE == this.visibility)
      return true;
    return false;
  }
  public IDManager getIDManager()
  {
    return parent.getIDManager();
  }

  public Sheet getParent()
  {
    return this.parent;
  }
  
  public int insertCoverInfo(CoverInfo ci)
  {
    Position pos = ModelHelper.insert(this.coverList, ci, getCoverInfoIndexMethod, repeatCoverInfoMethod);
    return pos.index;
  }

  public CoverInfo getCoverInfo(int index)
  {
    Position pos = ModelHelper.search(this.coverList, index, getCoverInfoIndexMethod, repeatCoverInfoMethod);
    if(pos.isFind)
      return this.coverList.get(pos.index);
    return null;
  }

  public List<CoverInfo> getCoverList()
  {
    return this.coverList;
  }
  
  @Override
  public boolean isMergable(BasicModel model)
  {
    int curSIndex = this.getIndex();
    int curEIndex = curSIndex + this.repeatedNum;
    
    int nSIndex = model.getIndex();
    int nEIndex = nSIndex + model.getRepeatedNum();
    
    if(curEIndex + 1 == nSIndex || nEIndex + 1 == curSIndex)
    {
      Column col = (Column) model;
      if(this.width != col.getWidth() 
          || this.visibility != col.getVisibility() 
          || this.style != col.getStyle())
        return false;
      return true;
    }  
    return false;
  }
  
  public String toString()
  {
    return new ToStringBuilder(this, ToStringStyle.SHORT_PREFIX_STYLE).append("index", getIndex()).append("repeat", getRepeatedNum()).toString();
  }

  @Override
  public void remove()
  {
    setStyle(null);
    parent = null;
  }
}

File: docs-web.resources/com.ibm.docs.web.resources/WebContent/js/wseditor/js/model/Column.js
/* ***************************************************************** */
/*                                                                   */
/* IBM Confidential                                                  */
/*                                                                   */
/* IBM Docs Source Materials                                         */
/*                                                                   */
/* (c) Copyright IBM Corporation 2012. All Rights Reserved.          */
/*                                                                   */
/* U.S. Government Users Restricted Rights: Use, duplication or      */
/* disclosure restricted by GSA ADP Schedule Contract with IBM Corp. */
/*                                                                   */
/* ***************************************************************** */

/****************************************************/
/* created by LinXin, linxinbj@cn.ibm.com           */
/*   1/27/2011                                      */
/* Column Model									    */
/****************************************************/
dojo.provide("websheet.model.Column");
dojo.require("websheet.model.BasicModel");
dojo.declare("websheet.model.Column",[websheet.model.BasicModel],{
	/*String*/_id : null,					//column id
	/*integer*/_width:null,					//column width (integer)
	/*String*/_styleId:null,				//column style id
	/*integer*/_repeatedNum:0,			    //column repeated number (integer)
	/*Sheet*/_parent: null,					//Sheet
	/*Document*/_doc: null,					//Document
	/*String*/_visibility:null,				//for visibility attr, if non-null, always invisible state
	/*Boolean*/_visible: true,				//hide or show column
	/**
	 * constructor
	 */
	constructor:function(parent,colId,width,styleId,repeatednum,visibility){
		this._id = colId;
		this._parent = parent;
		this._doc = this._parent._parent;
		if(undefined != styleId){
			this.setStyleId(styleId);
		}
		if(undefined != repeatednum){
			this._repeatedNum = repeatednum;
		}
		if(undefined !== width){
			this._width = width;
		}
		this._coverInfos = [];
		this.setVisibility(visibility);	
	},
	
	/*integer*/getCoverCellPosition:function(rowIndex,followStyle){
		var cells = this._coverInfos;
		var index = this._doc._mhelper.binarySearch(cells,rowIndex,this.equalCondition,"",followStyle,this._parent._id, websheet.Constant.Row);
		return index;
	},
	
	/**
	 * equal condition for binary search
	 * @param model					compare object
	 * @param index					it is a position for search
	 * @param fetchMethodById		how to fetch a object by id
	 * @param followStyle			if it is true,search a object whose style current position(parameter index) followed]
	 * 								otherwise only search a object in current position
	 * @param sheetId				sheet id
	 */
	/*integer*/equalCondition:function(model,index,fetchMethodById,followStyle,sheetId, methodType, idManager){
		var id = model._parent.getId();
		var repeatednum = model.getRowSpan() - 1;
		
		var pos = idManager.getRowIndexById(sheetId,id);
		
		if(pos != -1){pos += 1;}
		if(followStyle){
			if(pos	<= index && index <= pos + repeatednum){
				return 0;
			}else if(pos > index){
				return 1;
			}else{
				return -1;
			}
		}else{
			if(pos > index){
				return 1;
			}else if(pos == index){
				return 0;
			}else{
				return -1;
			}
		}
	},

	insertCoverInfo: function(coverInfo){
		var pos = this.getCoverCellPosition(coverInfo.getRow());
		if(pos <= -1){
			pos = -(pos + 1);
		} else {
			if(this._coverInfos[pos] == coverInfo)
				return pos;
		}
		
		this._coverInfos.splice(pos, 0, coverInfo);
	    return pos;
	},

	getCoverInfo: function(rowIndex, followStyle) {
		var index = this.getCoverCellPosition(rowIndex, followStyle);
		if(index >= 0)
			return this._coverInfos[index];
		return null;
	},
	
	/*
	 * check whether this object can be merged together with given column model 
	 */
	/*boolean*/isMergable:function(column) {
		if (column == null) return false;
		//repeated number attribute means that two column have the same style and width
		if(this.getWidth() != column.getWidth()){
			return false;
		}
		if(this.getStyleId() != column.getStyleId()){
			return false;
		}
		if(this.getVisibility() != column.getVisibility())
			return false;
		
		if(this._coverInfos.length > 0)
			return false;
		return true;
	},
	
	/**
	 * load data
	 */
	/*void*/load:function(colData){
	},
	/**
	 * get Sheet
	 */
	/*Sheet*/_getParent:function(){
		return this._parent;
	},
	/**
	 * get Document
	 */
	/*Document*/_getDocument:function(){
		return this._doc;
	},
	/**
	 * if column is null,return true
	 */
	/*boolean*/isNull:function(){
		if(this.getWidth() !== undefined || this.getStyleId() || (this._visibility && this._visibility != websheet.Constant.COLUMN_VISIBILITY_ATTR.SHOW)){
			return false;
		}
		return true;
	},
	/*boolean*/isVisible:function(){
		return this._visible;
	},
	/*String*/getVisibility: function(){
		if(this._visibility === null)
			return websheet.Constant.COLUMN_VISIBILITY_ATTR.SHOW;
		else
			return this._visibility;  
		 },
	/*void*/setVisibility:function(visibility){
		if(!visibility || visibility == websheet.Constant.COLUMN_VISIBILITY_ATTR.SHOW) 
			visibility = null;
		this._visibility = visibility;
		
		if(visibility == websheet.Constant.COLUMN_VISIBILITY_ATTR.HIDE){
		    this._visible = false;
		}else
			this._visible = true;
	},
//	/**
//	 * clear cells's value in this column
//	 */
//	/*void*/clearColumn:function(){
//		var sheet = this._parent;
//		var index = this.getIndex();
//		dojo.forEach(sheet.getRows(),function(row,rIndex){
//				var cell = row.getCell(index);
//				if(cell){
//					row.setCellRawValue("");
//				}
//				//if row is null,delete its instance in sheet
//				if(row.isNull()){
//					sheet.splice(rIndex,1);
//				}
//		});
//	},
	/**
	 * return column id
	 */
	/*string*/getId:function(){
		return this._id;
	},
	/**
	 * set column id
	 */
	/*void*/setId:function(id){
		this._id = id;
	},
	/**
	 * return column index, 1-based
	 */
	/*string*/getIndex:function(){
		var sheet = this._parent;
		var index = sheet.getColIndex(this._id);
		if(index > -1)
			index++;
		return index;
	},
	/**
	 * return column width
	 */
	/*integer*/getWidth: function() {
		return this._width;
	},

	/**
	 * set column width
	 * @param width     integer
	 */
	/*void*/setWidth:function(width){
		this._width = width;
	},
	/**
	 * return column repeated number
	 */
	/*integer*/getRepeatedNum: function() {
		return this._repeatedNum;
	},

	/**
	 * set column RepeatedNum
	 * @param repeatedNum     integer
	 */
	/*void*/setRepeatedNum:function(repeatedNum){
		this._repeatedNum = repeatedNum;
	},
	/**
	 * return column style
	 */
	/*Style*/getStyle:function(){
		var styleCode = null;
		if(this._styleId)
		{
			var styleManager = this._doc._getStyleManager();
			styleCode = styleManager.getStyleById(this._styleId);
		}
		return styleCode;
	},
	/**
	 * return column style id
	 */
	/*string*/getStyleId: function() {
	     return this._styleId;
	},
	/**
	 * merge style with style attribute
	 */
	/*void*/mergeStyle:function(styleAttr){
		//StyleManager merge style fetch by current style id with style attributes
		//return a new style id to cover original style id
		var styleId = this._doc._getStyleManager().changeStyle(this.getStyleId(),styleAttr);
		this._styleId = styleId;
	},
    /**
     * set column style id
     */
	/*void*/setStyleId:function(styleId){
		var styleManager = this._doc._getStyleManager();
		if(undefined != this._styleId){
			if(this._styleId == styleId){return;}
//			styleManager.deleteStyle(this._styleId);
		}
		this._styleId = styleId;
		if(styleId){
			styleManager.addRefCount(styleId);
		}
	}
});